File: projects/mamba/benches/lang/spectral_norm/spectral_norm.py
"""spectral-norm (Computer Language Benchmarks Game) — idiomatic pure Python.

Same algorithm and same N as the sibling spectral_norm.go, so python3 / mamba
(both run THIS file) and go (the .go) are a like-for-like comparison. No numpy:
the point is "idiomatic pure Python -> native", not "Python calling C".
"""

import math

N = 1000


def eval_A(i, j):
    return 1.0 / ((i + j) * (i + j + 1) // 2 + i + 1)


def eval_A_times_u(u):
    n = len(u)
    return [sum(eval_A(i, j) * u[j] for j in range(n)) for i in range(n)]


def eval_At_times_u(u):
    n = len(u)
    return [sum(eval_A(j, i) * u[j] for j in range(n)) for i in range(n)]


def eval_AtA_times_u(u):
    return eval_At_times_u(eval_A_times_u(u))


u = [1.0] * N
v = u
for _ in range(10):
    v = eval_AtA_times_u(u)
    u = eval_AtA_times_u(v)

vBv = sum(ue * ve for ue, ve in zip(u, v))
vv = sum(ve * ve for ve in v)
print("%.9f" % math.sqrt(vBv / vv))

File: apps/mamba/benches/lang/spectral_norm/spectral_norm.go
// spectral-norm (Computer Language Benchmarks Game) — idiomatic Go.
// Same algorithm and same N as the sibling spectral_norm.py, for a like-for-like
// python / go / mamba comparison. No unsafe / SIMD intrinsics — plain Go.
package main

import (
	"fmt"
	"math"
)

const N = 1000

func evalA(i, j int) float64 {
	return 1.0 / float64((i+j)*(i+j+1)/2+i+1)
}

func evalATimesU(u []float64) []float64 {
	n := len(u)
	v := make([]float64, n)
	for i := 0; i < n; i++ {
		var s float64
		for j := 0; j < n; j++ {
			s += evalA(i, j) * u[j]
		}
		v[i] = s
	}
	return v
}

func evalATTimesU(u []float64) []float64 {
	n := len(u)
	v := make([]float64, n)
	for i := 0; i < n; i++ {
		var s float64
		for j := 0; j < n; j++ {
			s += evalA(j, i) * u[j]
		}
		v[i] = s
	}
	return v
}

func evalAtATimesU(u []float64) []float64 {
	return evalATTimesU(evalATimesU(u))
}

func main() {
	u := make([]float64, N)
	for i := range u {
		u[i] = 1.0
	}
	v := u
	for k := 0; k < 10; k++ {
		v = evalAtATimesU(u)
		u = evalAtATimesU(v)
	}
	var vBv, vv float64
	for i := 0; i < N; i++ {
		vBv += u[i] * v[i]
		vv += v[i] * v[i]
	}
	fmt.Printf("%.9f\n", math.Sqrt(vBv/vv))
}
